Non-coplanar Lap joints using `Pocket`

papachap opened the pull request:
<img width="1919" alt="Screenshot 2025-02-18 163851" src="https://github.com/user-attachments/assets/6cc4def7-515e-4480-abeb-abc781b45c4a" />
<!-- Thank you for your pull request!  -->
<!-- Please start by describing your change in a few sentences. -->
<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### What type of change is this?
This is a rework of the Lap joints (X,L,T) to generate BTLx parameters when beams are not coplanar.
This is done by implementing and untilizing the `Pocket` BTLx Processing.
This update deprecates all remaining components of the old system.

- [ ] Bug fix in a **backwards-compatible** manner.
- [X] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.
- [ ] Other (e.g. doc update, configuration, etc)

### Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [X] I added a line to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading (e.g. `Added`, `Changed`, `Removed`).
- [X] I ran all tests on my computer and it's all green (i.e. `invoke test`).
- [X] I ran lint on my computer and there are no errors (i.e. `invoke lint`).
- [X] I added new functions/classes and made them available on a second-level import, e.g. `compas_timber.datastructures.Beam`.
- [X] I have added tests that prove my fix is effective or that my feature works.
- [X] I have added necessary documentation (if appropriate)

gonzalocasas reviewed (commented):
It looks very good to me, I am not familiar with the internals of the library, but I find the solution very readable and logical. I will test a bit locally and report back! Thanks a lot!!!

gonzalocasas commented:
I tested on a few different variations of the reciprocal frames we have for Coding Architecture II and I think it works well. The only odd situation I noticed is that there seems to be a validation that sometimes is not triggered.

The following code works as expected, and validates that the planes are not intersecting (you should be able to copy&paste the snippet as it is and it should reproduce exactly because I "internalized" all the data as compas objects):
```python
from compas.geometry import Line, Point
from compas_rhino.conversions import line_to_compas
from compas_timber.connections import XLapJoint
from compas_timber.elements import Beam
from compas_timber.model import TimberModel

line1 = Line(Point(x=0.368221844394343, y=-2.041176829546737, z=-0.33008107593548097), Point(x=11.321569321014982, y=-0.5060882451085971, z=0.49749674675594435))
line2 = Line(Point(x=1.719291946301213, y=2.4763282055645433, z=-0.5199851697376872), Point(x=3.1615173770327774, y=-5.356743352984839, z=1.954151291063396))
line3 = Line(Point(x=-0.4920312929518418, y=2.5050748635735807, z=-0.3364472635488296), Point(x=9.913185740559799, y=-4.617874680939579, z=0.09964174461654274))
lines = [line1, line2, line3]

width = 0.5
height = 0.3
z_vector = (0, 0, 1)

model = TimberModel()
for line in lines:
    beam = Beam.from_centerline(centerline=line, width=width, height=height, z_vector=z_vector)
    model.add_element(beam)

beams = list(model.beams)
XLapJoint.create(model, beams[0], beams[1], False, cut_plane_bias=0.25)
XLapJoint.create(model, beams[1], beams[2], False, cut_plane_bias=0.25)
XLapJoint.create(model, beams[2], beams[0], False, cut_plane_bias=0.25)

model.process_joinery()

geo = []
for beam in model.beams:
    geo.append(beam.geometry)

```
That results in this, which I think it is correct:
```
Traceback (most recent call last):
  File "rhinocode:///grasshopper/1/2a14e4d9-8ec6-42b4-a1eb-d2d2d16318c5/03461a0c-3d4a-47a3-b104-d4612fda66a6", line 36, in <module>
  File "/Users/gcasas/.rhinocode/py39-rh8/site-envs/ca-fs25-ooFiHkpI/compas_timber/model/model.py", line 347, in process_joinery
    joint.add_features()
  File "/Users/gcasas/.rhinocode/py39-rh8/site-envs/ca-fs25-ooFiHkpI/compas_timber/connections/x_lap.py", line 56, in add_features
    main_lap_feature = Pocket.from_volume_and_beam(negative_volume_main, self.main_beam, self.main_ref_side_index)
  File "/Users/gcasas/.rhinocode/py39-rh8/site-envs/ca-fs25-ooFiHkpI/compas_timber/fabrication/pocket.py", line 336, in from_volume_and_beam
    raise ValueError("The planes do not intersect. " + str(te))
ValueError: The planes do not intersect. compas.geometry.point.Point() argument after * must be an iterable, not NoneType
```

However, this code, in which one of the lines is a bit moved, but it's still not producing a joint, does not fail, and simply goes on without creating the 3rd lap joint:
```python
from compas.geometry import Line, Point
from compas_timber.connections import XLapJoint
from compas_timber.elements import Beam
from compas_timber.model import TimberModel

line1 = Line(Point(x=0.368221844394343, y=-2.041176829546737, z=-0.33008107593548097), Point(x=11.321569321014982, y=-0.5060882451085971, z=0.49749674675594435))
line2 = Line(Point(x=1.719291946301213, y=2.4763282055645433, z=-0.5199851697376872), Point(x=3.1615173770327774, y=-5.356743352984839, z=1.221114184647886))
line3 = Line(Point(x=-0.4920312929518418, y=2.5050748635735807, z=-0.3364472635488296), Point(x=9.913185740559799, y=-4.617874680939579, z=0.09964174461654274))
lines = [line1, line2, line3]

width = 0.5
height = 0.3
z_vector = (0, 0, 1)

model = TimberModel()
for line in lines:
    beam = Beam.from_centerline(centerline=line, width=width, height=height, z_vector=z_vector)
    model.add_element(beam)

beams = list(model.beams)
XLapJoint.create(model, beams[0], beams[1], False, cut_plane_bias=0.25)
XLapJoint.create(model, beams[1], beams[2], False, cut_plane_bias=0.25)
XLapJoint.create(model, beams[2], beams[0], False, cut_plane_bias=0.25)

model.process_joinery()

geo = []
for beam in model.beams:
    geo.append(beam.geometry)

```

The latter example results in this geometry, apparently skipping the validation that should indicate that the planes are not intersecting on one of the 3 joints:
![image](https://github.com/user-attachments/assets/ce52bbdd-2a1d-4a6e-abe4-faf1191856e9)


As I mentioned, I don't know the inner works of the library, so perhaps this is a correct outcome, but it feels slightly strange from a user PoV

gonzalocasas commented:
Noticed though that I have not tested the actually BTLx generation at all, I'm just verifying visually based on the output geometry!

gonzalocasas commented:
This I don't understand. Why is this beam cut short? If I moved it slightly in any direction, it goes back to normal.

![image](https://github.com/user-attachments/assets/22a516d7-0635-47bc-aac0-71ce6a5c2505)


```python

from compas_timber.connections import XLapJoint
from compas_timber.elements import Beam
from compas_timber.model import TimberModel
from compas.geometry import Line, Point
lines = [
    Line(Point(x=2.177033526832209, y=1.2437280126968124, z=0.050913702861557775), Point(x=4.9708346486781245, y=4.722273370190775, z=-0.03608410456129721)),
    Line(Point(x=4.08039057401605, y=4.885636545756687, z=0.0), Point(x=5.706171911685749, y=1.0691000335201553, z=0.0)),
    Line(Point(x=5.952401585817012, y=1.7806035147175316, z=0.06170623771252254), Point(x=1.3801981596614157, y=2.522672205860091, z=-0.11198141046628371)),
]
width = 0.1
height = 0.1
z_vector = (0, 0, 1)

model = TimberModel()
for line in lines:
    beam = Beam.from_centerline(centerline=line, width=width, height=height, z_vector=z_vector)
    model.add_element(beam)

beams = list(model.beams)
XLapJoint.create(model, beams[0], beams[1], False, cut_plane_bias=0.25)
XLapJoint.create(model, beams[1], beams[2], False, cut_plane_bias=0.25)
XLapJoint.create(model, beams[2], beams[0], False, cut_plane_bias=0.25)

model.process_joinery()

geo = []
for beam in model.beams:
    geo.append(beam.geometry)
```

papachap commented:
Hi @gonzalocasas! Thanks for testing! I checked the issue and just pushed a small adjustment to the code.

The issue was related to how the subtraction volumes for the `LapJoint` were generated and assigned to each element to create the `Pocket` feature. This fix ensures that the volumes are generated and assigned correctly.

Please run another round of testing. Keep in mind that flipping the lap side or using a large lap bias (especially when the centerlines are offset) might lead to errors if the volume does not intersect with the beam.

obucklin commented:
Hey @papachap  if you have time I would like to discuss this just to understand what you did.

papachap commented:
> Hey @papachap if you have time I would like to discuss this just to understand what you did.

sure, do you want to come up whenever you have time?

obucklin commented on src/compas_timber/fabrication/pocket.py:
```diff
@@ -0,0 +1,685 @@
+    ########################################################################
+
+    @classmethod
+    def from_volume_and_beam(cls, volume, beam, ref_side_index=0):
```
pocket is definitely applicable for plates, please change to `element`

obucklin replied:
This should also get the generic `from_shapes_and_elements()` method to make it compatible with `BTLxFromGeometryDefinition`. maybe that is where you solve for the `ref_face` and leave it as user-defined here.

obucklin commented on src/compas_timber/fabrication/pocket.py:
```diff
@@ -0,0 +1,685 @@
+        # cls._validate_volume_intersection(volume, beam, ref_side_index) # TODO: PluginNotInstalledError
+
+        # get ref_side, and ref_edge from the beam
+        ref_side = beam.ref_sides[ref_side_index]
```
I think this should solve for the correct `ref_side`.

obucklin commented on tests/compas_timber/test_pocket.py:
```diff
@@ -0,0 +1,327 @@
+        assert "Volume must have 6 faces." in str(e)
+
+
+# def test_pocket_with_non_intersecting_volume(beam):  # TODO: PluginNotInstalledError -> intersection_ray_mesh
```
remove?

papachap replied:
I think this is an important check because even if they don't intersect, it might still be able to generate parameters (although they want make any sense). And this is probably the less computational expensive way to check the intersection.
Nevertheless since it's not implemented yet I will remove it.

obucklin commented on src/compas_timber/fabrication/pocket.py:
```diff
@@ -0,0 +1,685 @@
+        # type: (Pocket) -> None
+        super(PocketParams, self).__init__(instance)
+
+    def as_dict(self):
```
due to change soon... Is there a way to mark this so that we remember to change it with the upcoming `Contour` refactoring?

papachap replied:
I have a feeling we will merge this PR before the `Contour` refactoring ;p
Anyhow I can add a ticket to the backlog

obucklin replied:
sounds good

obucklin commented on src/compas_timber/connections/lap_joint.py:
```diff
@@ -229,24 +167,24 @@ def _create_polyhedron(plane_a, lines, bias):  # Hexahedron from 2 Planes and 4
         # Get Beam Faces (Planes) in right order
-        planes_main = self._sort_beam_planes(beam_a, plane_cut_vector)
+        planes_main = self._sort_beam_planes(main_beam, plane_cut_vector)
         plane_a0, plane_a1, plane_a2, plane_a3 = planes_main
```
Please make a ticket to look at this again soon.

papachap replied:
are you referring to the way we generate the negative volumes?

obucklin replied:
yeah this is unclear and could use some refactoring

obucklin commented on src/compas_timber/connections/lap_joint.py:
```diff
@@ -264,6 +202,14 @@ def _create_negative_volumes(self):
-        negative_polyhedron_main_beam = self._create_polyhedron(plane_a0, lines, self.cut_plane_bias)
-        negative_polyhedron_cross_beam = self._create_polyhedron(plane_b0, lines, self.cut_plane_bias)
+        negative_polyhedron_main_beam = self._create_polyhedron(plane_b0, lines, self.cut_plane_bias)
+        negative_polyhedron_cross_beam = self._create_polyhedron(plane_a0, lines, self.cut_plane_bias)
```
I would like to see the changes we discussed about how the cut_plane is found implemented.

obucklin reviewed (changes requested):
hey i made a PR with suggested changes to generating parameters. please take a look and see if it is an improvement when applied to the CiA model. 

In my testing, the geometry was not generated correctly when the centerlines actually intersect. see this image:
![image](https://github.com/user-attachments/assets/9d2c24f1-e62b-4efd-b928-f7ef2696b201)

papachap commented:
@obucklin  Pushed some changes based on the comments. Meanwhile i will check the suggestions you made in the other PR.

1) followed a different approach for calculating the BTLx params (same offset issue in the BTL viewer).
2) made ref_side_index optional and implemented a method for getting the most optimal ref_side based on the intersections with the volume provided.
3) added `from shapes_and_element` class method.
4) renamed and upstreamed `angle_vectors_projected` to `compas_timber.utils`.

ps. opened backlog issues for the things we discussed.

papachap commented:
@obucklin @chenkasirer  This PR is ready for re-review. I already merged the side PR that was created during this review.

chenkasirer commented on src/compas_timber/fabrication/lap.py:
```diff
@@ -567,59 +801,22 @@ def volume_from_params_and_beam(self, beam):
         # ensure vertices are defined in counter-clockwise order when viewed from the outside
         faces = [
-            [3, 2, 1, 0],  # Bottom face
+            [0, 1, 2, 3],  # Bottom face
```
are these all reversed on purpose?

papachap replied:
Yes. I can't really remember what caused this change, but I can confirm that the geometry that is generated now is correct and valid.

chenkasirer commented on src/compas_timber/fabrication/pocket.py:
```diff
@@ -0,0 +1,742 @@
+    ########################################################################
+
+    @property
+    def params_dict(self):
```
I think we renamed this to `params` in a recent rework

chenkasirer commented on src/compas_timber/fabrication/pocket.py:
```diff
@@ -0,0 +1,742 @@
+        start_plane, end_plane, front_plane, back_plane, bottom_plane, top_plane = planes
+
+        machining_limits = MachiningLimits()
+        # machining_limits.face_limited_top = is_point_behind_plane(top_plane.point, ref_side)
```
remove?

chenkasirer commented on src/compas_timber/utils/__init__.py:
```diff
@@ -291,6 +292,32 @@ def distance_segment_segment(segment_a, segment_b):
     return distance_point_point(pt_seg_a, pt_seg_b)
 
 
+def angle_vectors_projected(vector_1, vector_2, normal):
```
how do you feel about naming these `vector_a` and `vector_b` instead? just for consistency with the other functions in this module.

chenkasirer commented on src/compas_timber/utils/__init__.py:
```diff
@@ -291,6 +292,32 @@ def distance_segment_segment(segment_a, segment_b):
+        The first vector.
+    vector_2 : :class:`compas.geometry.Vector`
+        The second vector.
+    normal : :class:`compas.geometry.Vector`
```
if `Plane` and `Frame` are also accepted, add them here

chenkasirer commented on src/compas_timber/utils/__init__.py:
```diff
@@ -291,6 +292,32 @@ def distance_segment_segment(segment_a, segment_b):
+    float
+        The angle between the two projected vectors
+    """
+    if isinstance(normal, Plane) or isinstance(normal, Frame):
```
```suggestion
    if isinstance(normal, (Plane, Frame)):
```

papachap commented on src/compas_timber/fabrication/pocket.py:
```diff
@@ -0,0 +1,742 @@
+        start_x, start_y, start_depth = cls._calculate_start_x_y_depth(ref_side, start_point)
+
+        # x"-axis and y"-axis (planar axis of the bottom face of the volume)
+        xxaxis = Vector.from_start_end(start_point, end_point)
+        yyaxis = Vector.from_start_end(start_point, back_point)
```
You are right pointing that out. I tried to follow the notation from the BTLx Documentation when naming the transformed 'prime' axes (x'-axis, x''-axis). But I understand that this might be confusing, though. We should either find a better way to represent the 'prime' vector or come up with a more descriptive name. I'm open to suggestions :)

papachap commented:
@chenkasirer I just pushed the requested changes and responded to some of your comments. This is ready for re-review

chenkasirer reviewed (approved):
LGTM

chenkasirer commented:
I think we should merge this and continue dealing with the various issues on `main`, don't think there's any inherent issues in this particular PR @papachap @gonzalocasas what do you guys think?
